### DISCUSSION TITLE
Getting data via getValues in functional component

### DISCUSSION THREAD

**Discussion by robertmarkevans** [react-querybuilder/react-querybuilder#800] [Q&A] [answered]
Had the Query Builder all working fine, pulling data feeds in from various API endpoints for fields, operators and values using a class based component
Just converted it to a functional component which was simple enough, but now the data via getValues is not populating the values dropdown for each Rule.
Has anyone got some working example code of a getValues function that uses Axios to await some data which is then returned as OptionList for the values choice for a Rule ??

**Comment by jakeboone02** [comment]
Is your `getValues` function `async`? That's currently unsupported, but being considered (see #775).

Can you share a code sample? I can show you how we asynchronously load value lists at my day job.

**Comment by robertmarkevans** [comment]
Hi Jake,

I read that Issue earlier and wasn't sure exactly what it meant as async.
What I think it meant was that as you type, the component will
asynchronously deliver some of the choices that start with what you've typed
This is not  what I am trying to do, it's much simpler than that. All I
have is a small function called getValues that uses Axios to get the data
for the values of the Rule, by being called by the getValues prop.
As I'm sure you're aware Axios returns a promise, so the function is
therefore async to allow me to await the data

Through console logs I can see that the data returns in the correct format
from Axios and that is being returned to the prop, but when you click on
the dropdown in the builder I get no options
Read also about getValueEditorType needed to return one of three values,
mine is 'select' ( however I'm unsure whether this is available when
getValues runs as it's not documented which order the props run (eg.
getOperators, getValues etc.. )

What's a mystery is the fact that as a class component it all works fine,
but the identical code with just the usual changes to make it functional
doesn't, but get so close !

Hope this makes sense.  Can provide code right now as I'm back home, and
code is all at work.

Rob

**Reply by jakeboone02** [reply]
Ok I think I understand. There are definitely different scenarios when it comes to "async"--autocomplete vs. lazy load, etc.--and I wasn't clear about that in #775 either.

That's a good point about the run order of function props, but I'm not sure it matters too much because they don't really depend on each other. You might try using the `fieldData` argument to get the value of other function props. For example, to get the `valueEditorType` from within the `getValues` function:

```ts
const getValues = (f, o, { fieldData }) => {
  const valueEditorType = typeof fieldData.valueEditorType === 'function'
    ? fieldData.valueEditorType(o)
    : fieldData.valueEditorType;
  return valueEditorType === 'select' ? [...selectValues] : [];
}
```

I would _love_ to see the diff between your class and function components when you get a chance (mostly just to see a class component again...feels like a lifetime since I've seen one of those!).

**Reply by robertmarkevans** [reply]
Replaced my getValues with your code snippet.  fieldData doesn't have a valueEditorType so getValues returns []
Is this my problem do you think, and if so, what's the fix ?
Regards

**Reply by robertmarkevans** [reply]
Here is a bare bones version. This code works, as clearly the getValues function is not async as it would need to be if we were using Axios. However, to replicate my problem just make getValues async. 
```
import React from "react"
import QueryBuilderProvider from "react-querybuilder"

export const QueryBuilder = () => {
  const [fields, setFields] = React.useState([
    { name: "field_1", label: "Field 1" },
    { name: "field_2", label: "Field 2" },
  ]);

  const getValueEditorType = () => "select";

  const getOperators = () => {
    return [
      { name: "=", label: "is"},
      { name: "contains", label: "contains"},
    ];
  };

  const getValues = () => [               --->  alternative non-working line   const getValues = async () => {
    { name: "value_1", label: "Value 1" },
    { name: "value_2", label: "Value 2" },
  ]);

  return (
    <QueryBuilderProvider
      fields={fields}
      getOperators={getOperators}
      getValueEditorType={getValueEditorType}
      getValues={getValues}
      enableDragAndDrop={false}
    />
  );
};

export default QueryBuilder;

  ```

**Comment by jakeboone02** [comment, marked as answer]
Support for async option list loading was added in [v8.13.0](https://github.com/react-querybuilder/react-querybuilder/releases/tag/v8.13.0).

Documentation here: https://react-querybuilder.js.org/docs/tips/async-option-lists
